File: flutter_planbook/android/app/src/main/kotlin/com/bapaws/planbook/widget/WidgetSettings.kt
package com.bapaws.planbook.widget

import android.content.Context
import android.content.SharedPreferences
import es.antonborri.home_widget.HomeWidgetPlugin
import org.json.JSONArray

/**
 * Widget 配置统一管理
 * 对应 iOS 端 WidgetSettings.swift
 */
object WidgetSettings {
    private const val PREFERENCES = "HomeWidgetPreferences"

    private const val KEY_SELECTED_PRIORITY = "widget_quadrant_selected_priority"
    private const val KEY_BACKGROUND_ASSET = "widget_background_asset"
    private const val KEY_WIDGET_THEME = "widget_theme"
    private const val KEY_LIGHT_COLOR_SCHEME = "__settings_light_color_scheme_key__"
    private const val KEY_DARK_COLOR_SCHEME = "__settings_dark_color_scheme_key__"
    private const val KEY_FILTER_MODE_PREFIX = "widget_filter_mode_"

    // 与 Flutter SettingsRepository.kSettingsQuadrantConfigs 保持一致
    private const val KEY_QUADRANT_CONFIGS = "widget_quadrant_configs"

    // 与 Flutter SettingsRepository.kSettingsTaskAutoNoteRulesKey 保持一致
    private const val KEY_TASK_AUTO_NOTE_RULES = "__settings_task_auto_note_rules_key__"

    /// pending 完成状态：widget 点击后写入，UI 立刻反馈；
    /// Flutter 端真正改完 DB 后清除这一项。
    /// 字典格式（JSON）：{"taskId": true/false}。
    private const val KEY_PENDING_COMPLETION_STATES = "widget_pending_completion_states"

    /**
     * 当前登录用户 ID 在 SharedPreferences 中的 key。
     * 必须与 Flutter 端 `packages/planbook_repository/lib/users/users_repository.dart`
     * 中的 `kUserId` 保持一致——Flutter 在登录 / 注销时通过 home_widget 写入 / 清除。
     */
    private const val KEY_CURRENT_USER_ID = "__supabase_user_id__"

    private fun getPrefs(context: Context): SharedPreferences {
        return HomeWidgetPlugin.getData(context)
    }

    /** 当前过滤模式（按 widget 实例隔离） */
    fun getFilterMode(context: Context, appWidgetId: Int): TaskFilterMode {
        val rawValue = getPrefs(context).getString("$KEY_FILTER_MODE_PREFIX$appWidgetId", null)
        return TaskFilterMode.fromRawValue(rawValue)
    }

    fun saveFilterMode(context: Context, appWidgetId: Int, mode: TaskFilterMode) {
        getPrefs(context).edit().putString("$KEY_FILTER_MODE_PREFIX$appWidgetId", mode.rawValue).apply()
    }

    /**
     * 当前登录的 Supabase 用户 ID；未登录或值为空字符串时返回 null。
     *
     * 与 Flutter 端 DatabaseTaskTodayApi / InboxApi / OverdueApi 的查询语义对齐：
     *   未登录 → 仅展示 `tasks.user_id IS NULL` 的任务
     *   已登录 → 仅展示 `tasks.user_id = ?` 的任务
     */
    fun getCurrentUserId(context: Context): String? {
        val raw = getPrefs(context).getString(KEY_CURRENT_USER_ID, null)?.trim()
        return if (raw.isNullOrEmpty()) null else raw
    }

    /** 当前选中的象限优先级（中号组件切换） */
    fun getSelectedPriority(context: Context): TaskPriority {
        val rawValue = getPrefs(context).getString(KEY_SELECTED_PRIORITY, null)
        return TaskPriority.fromRawValue(rawValue)
    }

    fun setSelectedPriority(context: Context, priority: TaskPriority) {
        getPrefs(context).edit().putString(KEY_SELECTED_PRIORITY, priority.rawValue).apply()
    }

    /**
     * 四象限自定义名称（仅包含非空名称）。
     * Flutter 端 jsonEncode 写入 [{"priority","name"}, ...]。
     */
    fun getQuadrantConfigs(context: Context): Map<String, String> {
        val json = getPrefs(context).getString(KEY_QUADRANT_CONFIGS, null) ?: return emptyMap()
        return try {
            val arr = JSONArray(json)
            val result = mutableMapOf<String, String>()
            for (i in 0 until arr.length()) {
                val obj = arr.optJSONObject(i) ?: continue
                val priority = obj.optString("priority", "")
                if (priority.isEmpty()) continue
                val name = if (obj.isNull("name")) null else obj.optString("name", null)
                if (!name.isNullOrEmpty()) {
                    result[priority] = name
                }
            }
            result
        } catch (_: Exception) {
            emptyMap()
        }
    }

    /** 读取某个象限的自定义名称；未配置或为空时返回 null */
    fun getCustomQuadrantName(context: Context, priority: TaskPriority): String? {
        return getQuadrantConfigs(context)[priority.rawValue]
    }

    /** 背景资源名称（Flutter 同步写入） */
    fun getBackgroundAsset(context: Context): String {
        return getPrefs(context).getString(KEY_BACKGROUND_ASSET, "bg_dot") ?: "bg_dot"
    }

    /** 是否为深色模式（从 widget_theme 解析） */
    fun isDarkMode(context: Context): Boolean {
        val themeJson = getPrefs(context).getString(KEY_WIDGET_THEME, null) ?: return false
        return try {
            val theme = org.json.JSONObject(themeJson)
            theme.optBoolean("isDarkMode", false)
        } catch (_: Exception) {
            false
        }
    }

    /** 亮色主题颜色方案 JSON */
    fun getLightColorSchemeJSON(context: Context): String? {
        return getPrefs(context).getString(KEY_LIGHT_COLOR_SCHEME, null)
    }

    /** 暗色主题颜色方案 JSON */
    fun getDarkColorSchemeJSON(context: Context): String? {
        return getPrefs(context).getString(KEY_DARK_COLOR_SCHEME, null)
    }

    // ============================================
    // Pending 完成状态（Widget 乐观 UI）
    //
    // 工作流：
    //   1. 用户在 widget 上点击任务行 → WidgetClickReceiver 把翻转后的目标状态
    //      写到 pending map 并刷新 widget。
    //   2. Provider 在拼装 RemoteViews 时把 pending 覆盖到 DB 查到的 isCompleted，
    //      让 widget UI 立刻反馈（哪怕主 App 还没启动）。
    //   3. Flutter 端跑完业务后 WidgetClickReceiver 清除对应 taskId 的 pending，
    //      之后 widget 重新刷新就读到权威 DB 状态，pending 不再生效。
    // ============================================

    /** 读取所有 pending 状态 */
    fun getPendingCompletions(context: Context): Map<String, Boolean> {
        val json = getPrefs(context).getString(KEY_PENDING_COMPLETION_STATES, null) ?: return emptyMap()
        return try {
            val obj = org.json.JSONObject(json)
            val result = mutableMapOf<String, Boolean>()
            obj.keys().forEach { key ->
                result[key] = obj.optBoolean(key, false)
            }
            result
        } catch (_: Exception) {
            emptyMap()
        }
    }

    /** 写入某个任务的 pending 状态 */
    fun setPendingCompletion(context: Context, taskId: String, completed: Boolean) {
        val map = getPendingCompletions(context).toMutableMap()
        map[taskId] = completed
        val obj = org.json.JSONObject()
        map.forEach { (k, v) -> obj.put(k, v) }
        getPrefs(context).edit().putString(KEY_PENDING_COMPLETION_STATES, obj.toString()).apply()
    }

    /** 清除某个任务的 pending 状态 */
    fun clearPendingCompletion(context: Context, taskId: String) {
        val map = getPendingCompletions(context).toMutableMap()
        if (map.remove(taskId) == null) return
        if (map.isEmpty()) {
            getPrefs(context).edit().remove(KEY_PENDING_COMPLETION_STATES).apply()
        } else {
            val obj = org.json.JSONObject()
            map.forEach { (k, v) -> obj.put(k, v) }
            getPrefs(context).edit().putString(KEY_PENDING_COMPLETION_STATES, obj.toString()).apply()
        }
    }

    /** 获取单个任务的 pending 状态；不存在返回 null */
    fun pendingCompletion(context: Context, taskId: String): Boolean? {
        return getPendingCompletions(context)[taskId]
    }

    // ============================================
    // 任务自动创建笔记规则
    //
    // 规则在 Flutter 端用 jsonEncode 写入：
    // [
    //   { "type": "create"|"edit"|"createAndEdit"|"none",
    //     "tag": {...} | null,
    //     "priority": 1|2|3|4 | null,   // value，与 TaskPriority 的 value 对齐
    //     "is_subtask": true | false }
    // ]
    //
    // priority value 对照（见 packages/planbook_api/.../task_priority.dart）：
    //   1 = high, 2 = medium, 3 = low, 4 = none
    //
    // 小部件端无法弹出"编辑笔记"页面，所以 edit / createAndEdit 都退化为 create。
    // ============================================

    enum class TaskAutoNoteType {
        NONE, CREATE, EDIT, CREATE_AND_EDIT;

        val shouldCreate: Boolean
            get() = this == CREATE || this == CREATE_AND_EDIT
    }

    /**
     * 根据任务的 priority（数据库里的字符串 high/medium/low/none）和 isSubtask
     * 找到匹配的自动笔记规则类型；找不到时返回 NONE。
     */
    fun getTaskAutoNoteType(
        context: Context,
        taskPriorityRaw: String?,
        isSubtask: Boolean,
    ): TaskAutoNoteType {
        val json = getPrefs(context).getString(KEY_TASK_AUTO_NOTE_RULES, null) ?: return TaskAutoNoteType.NONE
        return try {
            val arr = JSONArray(json)
            for (i in 0 until arr.length()) {
                val rule = arr.optJSONObject(i) ?: continue
                val typeName = rule.optString("type", "none")
                val matches = if (isSubtask) {
                    rule.optBoolean("is_subtask", false)
                } else {
                    val priorityValue = if (rule.isNull("priority")) null else rule.optInt("priority")
                    !rule.optBoolean("is_subtask", false) &&
                        priorityValue != null &&
                        priorityValueToRaw(priorityValue) == (taskPriorityRaw ?: "none")
                }
                if (matches) {
                    return when (typeName) {
                        "create" -> TaskAutoNoteType.CREATE
                        "edit" -> TaskAutoNoteType.EDIT
                        "createAndEdit" -> TaskAutoNoteType.CREATE_AND_EDIT
                        else -> TaskAutoNoteType.NONE
                    }
                }
            }
            TaskAutoNoteType.NONE
        } catch (_: Exception) {
            TaskAutoNoteType.NONE
        }
    }

    private fun priorityValueToRaw(value: Int): String = when (value) {
        1 -> "high"
        2 -> "medium"
        3 -> "low"
        else -> "none"
    }

    /** 获取当前主题颜色方案 */
    fun getCurrentColorScheme(context: Context): FlutterColorScheme {
        val isSystemDark = context.resources.configuration.uiMode and
                android.content.res.Configuration.UI_MODE_NIGHT_MASK ==
                android.content.res.Configuration.UI_MODE_NIGHT_YES
        val isDark = isSystemDark || isDarkMode(context)
        val json = if (isDark) getDarkColorSchemeJSON(context) else getLightColorSchemeJSON(context)
        return try {
            json?.let { AppFlutterColorSchemes.decode(it) }
                ?: if (isDark) AppFlutterColorSchemes.redDark else AppFlutterColorSchemes.redLight
        } catch (_: Exception) {
            if (isDark) AppFlutterColorSchemes.redDark else AppFlutterColorSchemes.redLight
        }
    }
}

File: flutter_planbook/ios/Widgets/Core/WidgetSettings.swift
//
//  WidgetSettings.swift
//  Widgets
//
//  Widget 配置统一管理
//  所有通过 UserDefaults (App Group) 持久化的配置项，统一封装为类型安全的计算属性
//

import Foundation

@available(iOS 16.0, *)
enum WidgetSettings {
    private static var defaults: UserDefaults { .appGroup }

    // MARK: - Keys

    private enum Keys {
        static let selectedPriority = "widget_quadrant_selected_priority"
        static let backgroundAsset = "widget_background_asset"
        static let widgetTheme = "widget_theme"
        static let lightColorScheme = "__settings_light_color_scheme_key__"
        static let darkColorScheme = "__settings_dark_color_scheme_key__"
        /// 四象限自定义配置（名称），Flutter 端通过 home_widget 写入，
        /// 与 `SettingsRepository.kSettingsQuadrantConfigs` 保持一致。
        static let quadrantConfigs = "widget_quadrant_configs"
        /// pending 完成状态：widget AppIntent 翻转后写入这里，UI 立刻反馈；
        /// Flutter 端真正改完 DB 后会通过 MethodChannel 清除这一项。
        /// 字典格式：[taskId: completed]。
        static let pendingCompletionStates = "widget_pending_completion_states"
        /// 当前登录用户 ID。Flutter 端登录 / 注销时通过 home_widget 写入 / 清除，
        /// 必须与 `packages/planbook_repository/lib/users/users_repository.dart` 中
        /// 的 `kUserId` 常量保持一致。
        static let currentUserId = "__supabase_user_id__"
        /// 会员状态。Flutter 端通过 home_widget 写入；缺省视为非会员。
        static let isPremium = "widget_is_premium"
    }

    // MARK: - 象限优先级

    /// 当前选中的象限优先级（中号组件切换）
    static var selectedPriority: TaskPriority {
        get {
            guard let rawValue = defaults.string(forKey: Keys.selectedPriority),
                  let priority = TaskPriority(rawValue: rawValue) else {
                return .high
            }
            return priority
        }
        set {
            defaults.set(newValue.rawValue, forKey: Keys.selectedPriority)
        }
    }

    // MARK: - 背景

    /// 背景资源名称（Flutter 同步写入）
    static var backgroundAsset: String {
        defaults.string(forKey: Keys.backgroundAsset) ?? "bg_dot"
    }

    /// 是否为深色模式（从 widget_theme 解析）
    static var isDarkMode: Bool {
        guard let themeJson = defaults.string(forKey: Keys.widgetTheme),
              let data = themeJson.data(using: .utf8),
              let theme = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return false
        }
        return theme["isDarkMode"] as? Bool ?? false
    }

    // MARK: - 四象限自定义配置

    /// 四象限自定义名称（仅包含非空名称）
    static var quadrantConfigs: [String: String] {
        guard let json = defaults.string(forKey: Keys.quadrantConfigs),
              let data = json.data(using: .utf8),
              let array = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return [:]
        }
        var result: [String: String] = [:]
        for item in array {
            guard let priority = item["priority"] as? String,
                  let name = item["name"] as? String,
                  !name.isEmpty else { continue }
            result[priority] = name
        }
        return result
    }

    /// 读取某个象限的自定义名称；未配置或为空时返回 nil
    static func customQuadrantName(for priority: TaskPriority) -> String? {
        quadrantConfigs[priority.rawValue]
    }

    // MARK: - 主题颜色

    /// 亮色主题颜色方案 JSON
    static var lightColorSchemeJSON: String? {
        defaults.string(forKey: Keys.lightColorScheme)
    }

    /// 暗色主题颜色方案 JSON
    static var darkColorSchemeJSON: String? {
        defaults.string(forKey: Keys.darkColorScheme)
    }

    // MARK: - Pending 完成状态（Widget 乐观 UI）
    //
    // 工作流：
    //   1. 用户在 widget 上勾选任务 → `CompleteTaskIntent` 把翻转后的目标状态
    //      写到 `pendingCompletionStates[taskId]` 并 reload timeline。
    //   2. Provider 在拼装 entry 时把 pending 覆盖到 DB 查到的 `isCompleted`，
    //      让 widget UI 立刻反馈（哪怕主 App 还没启动）。
    //   3. AppIntent 链到主 App 跑 Flutter 完整业务，DB / Supabase 写完后
    //      Flutter 通过 MethodChannel 调过来清除该 taskId 的 pending，
    //      之后 widget 重新刷新就读到权威 DB 状态，pending 不再生效。
    //
    // 注意：写 DB 的责任完全在 Flutter 端，widget 进程不再直接改 DB。

    /// 取出当前所有 pending 状态。值为 `[taskId: completed]`。
    static func allPendingCompletions() -> [String: Bool] {
        guard let raw = defaults.dictionary(forKey: Keys.pendingCompletionStates) else {
            return [:]
        }
        var result: [String: Bool] = [:]
        for (key, value) in raw {
            if let bool = value as? Bool {
                result[key] = bool
            }
        }
        return result
    }

    /// 读取单个任务的 pending 状态；不存在返回 nil。
    static func pendingCompletion(forTaskId taskId: String) -> Bool? {
        guard let dict = defaults.dictionary(forKey: Keys.pendingCompletionStates),
              let value = dict[taskId] as? Bool else {
            return nil
        }
        return value
    }

    /// 写入或更新某个任务的 pending 状态。
    static func setPendingCompletion(taskId: String, completed: Bool) {
        var dict = defaults.dictionary(forKey: Keys.pendingCompletionStates) ?? [:]
        dict[taskId] = completed
        defaults.set(dict, forKey: Keys.pendingCompletionStates)
    }

    /// 清除某个任务的 pending 状态（Flutter 端 DB 写完后调用）。
    static func clearPendingCompletion(forTaskId taskId: String) {
        var dict = defaults.dictionary(forKey: Keys.pendingCompletionStates) ?? [:]
        guard dict.removeValue(forKey: taskId) != nil else { return }
        if dict.isEmpty {
            defaults.removeObject(forKey: Keys.pendingCompletionStates)
        } else {
            defaults.set(dict, forKey: Keys.pendingCompletionStates)
        }
    }

    // MARK: - 当前登录用户

    /// 当前登录的 Supabase 用户 ID，未登录或值为空时返回 nil。
    ///
    /// widget 进程需要据此把 SQL 过滤拼成 `tasks.user_id IS NULL`（未登录）或
    /// `tasks.user_id = ?`（已登录），与 Flutter 端 DatabaseTaskTodayApi /
    /// InboxApi / OverdueApi 的查询语义对齐，避免切换账号后看到上一个用户的数据。
    static var currentUserId: String? {
        let raw = defaults.string(forKey: Keys.currentUserId)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard let raw, !raw.isEmpty else { return nil }
        return raw
    }

    // MARK: - 会员

    /// 时间块小组件是否按会员渲染。从未写入时视为非会员，避免泄露日程。
    ///
    /// Flutter `home_widget` 经 StandardMessageCodec 写入的 Bool 在
    /// UserDefaults 里是 `NSNumber`，`as? Bool` 会失败，必须用 `bool(forKey:)`。
    static var isPremium: Bool {
        defaults.bool(forKey: Keys.isPremium)
    }
}
